orangecontrib/text/vectorization/dense.py

```diff
@@ -0,0 +1,164 @@
+from Orange.misc.server_embedder import ServerEmbedderCommunicator
+from orangecontrib.text import Corpus
+
+import zlib
+import base64
+import json
+import sys
+import numpy as np
+import warnings
+
+from typing import List, Tuple, Any, Optional
+
+AGGREGATORS = ['mean', 'sum', 'max', 'min']
+LANGS_TO_ISO = {'English': 'en', 'Slovenian': 'sl', 'German': 'de'}
+LANGUAGES = list(LANGS_TO_ISO.values())
+EMB_DIM = 300
+
+class PretrainedEmbedder:
+    """This class is used for obtaining dense embeddings of documents in 
+    corpus using fastText pretrained models from:
+    
+    E. Grave, P. Bojanowski, P. Gupta, A. Joulin, T. Mikolov, 
+    Learning Word Vectors for 157 Languages.
+    Proceedings of the International Conference on Language Resources and Evaluation, 2018.
+
+    Embedding is performed on server so the internet connection is a 
+    prerequisite for using the class. Currently supported languages are:
+        - English (en)
+        - Slovenian (sl)
+        - German (de)
+
+    Attributes:
+        language (str): ISO 639-1 (two-letter) code of desired language.
+        aggregator (str): Aggregator which creates document embedding (single 
+                          vector) from word embeddings (multiple vectors). 
+                          Allowed values are mean, sum, max, min.
+    """
+
+    def __init__(self, language: str = 'en',
+                 aggregator: str = 'mean') -> None:
+        
+        lang_error = '{} is not a valid language. Allowed values: {}'
+        agg_error = '{} is not a valid aggregator. Allowed values: {}'
+        if(language.lower() not in LANGUAGES):
+            raise ValueError(lang_error.format(language, ', '.join(LANGUAGES)))
+        else:
+            self.language = language.lower() 
+
+        if(aggregator.lower() not in AGGREGATORS):
+            raise ValueError(agg_error.format(aggregator, ', '.join(AGGREGATORS)))
+        else:
+            self.aggregator = aggregator.lower()
+
+        self._dim = EMB_DIM
```

Comment:
Before I didn't explain well enough: I would remove self._dim completely. In __call__ function I would build arrays based on the size of embeddings. So if you get embeddings that has 300 values for each document use this size in places where you would use self._dim before. Is it possible or I missed something here?

Reply:
Completely removing this parameter might cause problems in case of partially successful result. Unsuccessful embedding is represented as None instead of vector of nans so it would sometimes be impossible to determine dimension from first element in list. In such cases we would have to look for first successful embedding, determine dimension and then go back to replace None values with vectors of nans.

Reply:
I would say that determining shape from first non-None embedding should be better (it can be done before the loop). Why do I think so? Let say one day we will want to add another embedder which will have different dimension and then we will need to change the script because this value is hardcoded. I think it is better just to get it from data. 

When we are talking about this I taught about another case. In image-analytics, we do not put non-embedded images on same output (we have a separate output there). But there is another case -- some images get skipped because some embeddings cannot embed all type of images. In the document embedding the case is a bit different, skipped documents are most likely some embedding errors which need to be discovered and possibly fixed. I would actually suggest skipping documents that are not embedded at the output and report about that as a warning in the widget (when the length of input and output corpus differ). What do you think about that solution?

Reply:
I think your suggestion is probably the best possible solution. Returning vectors of nans isn't particularly useful anyway.